### PR TITLE
BUG: to_datetime with mixed-string-and-numeric

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Surfaced while implementing #55564

### CODE REVIEW

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -700,15 +700,15 @@ def array_to_datetime_with_tz(ndarray values, tzinfo tz, NPY_DATETIMEUNIT creso)
             ival = NPY_NAT
 
         else:
-            ts = Timestamp(item)
+            if PyDateTime_Check(item) and item.tzinfo is not None:
+                # We can't call Timestamp constructor with a tz arg, have to
+                #  do 2-step
+                ts = Timestamp(item).tz_convert(tz)
+            else:
+                ts = Timestamp(item, tz=tz)
```
Comment (mroeschke) [inline pandas/_libs/tslib.pyx:708]: Nit: Should we use `tz_localize` here to be more explicit?

Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:708]: the point is to match the behavior of calling the constructor this way, so i prefer this.

But now that #55712 is merged we can replace all of this with `convert_to_tsobject` which will move the ball down the field on a bunch of fronts.  will update

Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:708]: looks like moving to convert_to_tsobject introduces its own set of issues, so ill do that in a separate PR

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Comment by mroeschke** [comment]
@jbrockmendel mind taking a look at this test failure on main? This PR _might_ have been responsible https://github.com/pandas-dev/pandas/actions/runs/6728530133/job/18288031590

**Comment by jbrockmendel** [comment]
will do

**Comment by mroeschke** [comment]
Thanks!

**Comment by jbrockmendel** [comment]
i dont think a measured-in-minutes fix is forthcoming.  might need to xfail or comment-out that offending line

**Pull request merged**